gitcmd/gitcmd.go

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Package gitcmd contains utilities for common Git operations in a local repository, including
+// authentication with a remote repository.
+package gitcmd
+
+import (
+	"fmt"
+	"strings"
+)
+
+const githubPrefix = "https://github.com/"
+const azdoDncengPrefix = "https://dnceng@dev.azure.com/"
+
+// URLAuther manipulates a Git repository URL (GitHub, AzDO, ...) such that Git commands taking a
+// remote will work with the URL. This is intentionally vague: it could add an access token into the
+// URL, or it could simply make the URL compatible with environmental auth on the machine (SSH).
+type URLAuther interface {
+	// InsertAuth inserts authentication into the URL and returns it, or if the auther doesn't
+	// apply, returns the url without any modifications.
+	InsertAuth(url string) string
+}
+
+// GitHubSSHAuther turns an https-style GitHub URL into an SSH-style GitHub URL.
+type GitHubSSHAuther struct{}
```

Comment:
Given that we removed the `Git` prefix from `GitURLAuther` I would have expected that it is removed here too. Is there a reason it was kept?

Reply:
Well, here it isn't a Git prefix, rather a GitHub prefix. `HubSSHAuther` is confusing, and it is GitHub specific (other services don't define SSH URLs the same way, if they have them at all).

Reply:
makes sense